compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -541,8 +541,45 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                     self.requires_unsafe(expr.span, DerefOfRawPointer);
                 }
             }
-            ExprKind::InlineAsm { .. } => {
+            ExprKind::InlineAsm(box InlineAsmExpr {
+                asm_macro: _,
+                ref operands,
+                template: _,
+                options: _,
+                line_spans: _,
+            }) => {
                 self.requires_unsafe(expr.span, UseOfInlineAssembly);
+
+                // For inline asm, do not use `walk_expr`, since we want to handle the label block
+                // specially.
+                for op in &**operands {
+                    use rustc_middle::thir::InlineAsmOperand::*;
+                    match op {
+                        In { expr, reg: _ }
+                        | Out { expr: Some(expr), reg: _, late: _ }
+                        | InOut { expr, reg: _, late: _ } => self.visit_expr(&self.thir()[*expr]),
+                        SplitInOut { in_expr, out_expr, reg: _, late: _ } => {
+                            self.visit_expr(&self.thir()[*in_expr]);
+                            if let Some(out_expr) = out_expr {
+                                self.visit_expr(&self.thir()[*out_expr]);
+                            }
+                        }
+                        Out { expr: None, reg: _, late: _ }
+                        | Const { value: _, span: _ }
+                        | SymFn { value: _, span: _ }
+                        | SymStatic { def_id: _ } => {}
```

Comment:
The default visitor uses `visit_anon_const` and `visit_qpath` in this place, don't you need them here?

To avoid potential mistakes, it probably makes sense to add a public `fn walk_inline_asm_operand` helper to intravisit.rs.
```rust
match op {
  Label { .. } => self.in_safety_context(SafetyContext::Safe, walk_inline_asm_operand(...)),
  _ => walk_inline_asm_operand(...),
}
```

Reply:
This is the THIR visitor which doesn't have visit_anon_const and visit_qpath.

Reply:
ping @compiler-errors

Reply:
Sorry, pinged wrong person. @petrochenkov

Reply:
Sorry, I never noticed the update, it's better to set the `S-waiting-on-review` label if a PR is waiting on review.

If this part cannot be improved, then it's good to go.